### PR TITLE
PowerDNS dns provider.

### PR DESCRIPTION
**Pull request by kylejohnson**
Thank you for contributing to sewer.                    
Every contribution to sewer is important to us.                       
You may not know it, but you have just contributed to making the world a more safer and secure place.                         
           

## What(What have you changed?)

Added a dns provider for the PowerDNS API.

## Why(Why did you change it?)

So that DNS authentication can be performed against a PowerDNS server.

### CODE REVIEW

**Comment by komuw** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- sewer/dns_providers/powerdns.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5147628&bid=16908148)

File: sewer/dns_providers/powerdns.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                        {
+                            "content": f'"{domain_dns_value}"',
+                            "disabled": False
+                        }
+                    ]
+                }
+            ]
+        }
+
+        powerdns_response = requests.patch(
```
Comment (komuw) [inline sewer/dns_providers/powerdns.py]: Codacy Issue found: [Unused variable 'powerdns_response'](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5147628)

**Comment by komuw** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- sewer/dns_providers/powerdns.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5147628&bid=16908148)

**Comment by komuw** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- sewer/dns_providers/powerdns.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5147628&bid=16908148)

**Comment by komuw** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- sewer/dns_providers/powerdns.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5147628&bid=16908148)

**Comment by mmaney** [comment]
@kylejohnson I've had a need to know what the _domain.tld_ of the fqdn that's passed to the provider class is in a still-private hack for updating an acme-validation subzone (challenges are in host.acme-validation.domain.tld linked by static cnames in the apex zone).  One reason that provider is still private code is that it currently has the _domain.tld_ hardwired.  :-(

Would having that apex zone name passed to the dns code help with PowerDNS, or is that just one part of the search you have to do?  (or @anyone, would it be useful for other dns_providers?)

**Comment by kylejohnson** [comment]
@mmaney 

> Would having that apex zone name passed to the dns code help with PowerDNS

Yes, but I don't know that we ever really know the apex domain; you can request a certificate for any number of subdomains without also requesting for the apex domain, and you can also request certificates for multiple, separate apex domains (I think?).

Either way, PowerDNS makes it very easy to check if a given `zone_id` exists, and the [validate_powerdns_zone](https://github.com/komuw/sewer/pull/147/files#diff-df9d2c034c57a1ccdb9a88c0a020e80bR29) function takes care of the rest.  I imagine the code would not be applicable with different providers.

**Comment by komuw** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- sewer/dns_providers/powerdns.py  4
- sewer/dns_providers/tests/test_powerdns.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5147628&bid=16908148)

**Comment by kylejohnson** [comment]
So obviously that doesn't look good, but it looks like the testing for `sewer` isn't so great to begin with?

**Comment by mmaney** [comment]
@kylejohnson Ah, I see, you're using a part of the PowerDNS system to find the apex for each request.  I don't know if there was anything like that that I could use, but in any case I was coming from the perspective that I had a bunch of fqdns to validate (and they did share an apex, whether that was one of the fqdns or not), and of course I knew - the literal me setting up and running the requests - what the apex was.  If I'd had something like that virtual URI tree (?) that this driver queries... nah, at this point it would still be hard-wired, but I might be thinking of finding it in a like manner as a way forward.

Thanks for the food for thought!

**Comment by komuw** [comment]
@kylejohnson tests are failing

**Comment by komuw** [comment]
note to self; related: https://github.com/komuw/sewer/pull/144

**Comment by kylejohnson** [comment]
@komuw Sorry, didn't see the other refactoring changes included.  This PR now only has the powerdns changes.

**Comment by komuw** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- sewer/dns_providers/tests/test_powerdns.py  2
- sewer/dns_providers/powerdns.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5147628&bid=16908148)

**Comment by kylejohnson** [comment]
@komuw Tests for my changes look OK, I think.
Locally, they seem to run OK:
```
kjohnson@kyle-workstation01:~/work/sewer$ python -m unittest sewer.dns_providers.tests.test_powerdns.TestPowerDNS
...
----------------------------------------------------------------------
Ran 3 tests in 0.001s

OK
```

**Comment by komuw** [comment]
@kylejohnson 

You need to add powerdns to:     
    
1. `setup.py`; somewhere here: https://github.com/komuw/sewer/blob/ee0f259c303038968556d8b5680e6c58f0fd1645/setup.py#L29   
and also here: https://github.com/komuw/sewer/blob/ee0f259c303038968556d8b5680e6c58f0fd1645/setup.py#L111       
    
2. `https://github.com/komuw/sewer/blob/master/sewer/cli.py` somewhere here: https://github.com/komuw/sewer/blob/ee0f259c303038968556d8b5680e6c58f0fd1645/sewer/cli.py#L74   
and somehere here: https://github.com/komuw/sewer/blob/ee0f259c303038968556d8b5680e6c58f0fd1645/sewer/cli.py#L300-L311    
    
3. `README.md` somewhere here: https://github.com/komuw/sewer/blame/master/README.md#L26   
and somewhere here: https://github.com/komuw/sewer/blame/master/README.md#L70-L71

**Comment by komuw** [comment]
@kylejohnson the CI(circleCI) is failing because this PR is going to reduce the test code coverage to less than 85%. 
Screenshot 2020-03-23 at 06 05 40
 

As you can see, this is because the file `sewer/dns_providers/powerdns.py` only has a test coverage of 29%.   
Compare this to the file, `sewer/dns_providers/route53.py` which has a test coverage of 89%.    

The main lines in `sewer/dns_providers/powerdns.py` that are missing test coverage are line 55 to line 91:
Screenshot 2020-03-23 at 06 07 45    
https://github.com/komuw/sewer/blob/7fbcbccfdb56753aac8aed1c80c2810c88191971/sewer/dns_providers/powerdns.py#L55-L91    

So it seems like adding a test(unit test or otherwise) for the method `_common_dns_record` would help to restore the test code coverage

**Comment by komuw** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- sewer/dns_providers/powerdns.py  4
- sewer/dns_providers/tests/test_powerdns.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5147628&bid=16908148)

**Comment by kylejohnson** [comment]
@komuw Thank you for walking me through this.

**Comment by mmaney** [comment]
@kylejohnson I'm trying to sort out how best to solve #162, and powerdns is one of two providers that do not strip the "*." from wildcard domains.  Is that required for powerdns to work, or is it an oversight that will cause it to fail, like route53 does in that bug?

If powerdns doesn't need it, then the obvious fix is to not add it in sewer, and then the other providers can be cleaned up at our leisure.  If powerdns does need the star, then... Well, it's easy enough to patch route53 to strip it, as most of the providers probably do.  But I really hope we can clean this up instead.

Thanks!

**Comment by kylejohnson** [comment]
@mmaney Hmm, I didn't explicitly run across this in my tests, integrations or usage.  In fact, I haven't used sewer with a wildcard cert yet.

That said, it looks like powerdns is not happy:
`"message": "Error generating certificate - Error creating or deleting PowerDNS record: {\"error\": \"Name '_acme-challenge.*.domain.tld.' contains unsupported characters\"}"`

It looks like `validate_powerdns_zone` works correctly as it is able to walk backwards from `*.domain.tld` to find `domain.tld` as the apex zone.

I'll need to dig a little bit more and see what the exact query that powerdns.py is sending to the powerdns API, because for example I do have existing `CNAME` records for widlcard domains.

(Note that to produce the above error, I tried using a wildcard alt domain, e.g. `domain_name=domain, domain_alt_names=[f"*.{domain}"]`)

**Comment by mmaney** [comment]
Yes, that's the case which seems to trip up a couple of service providers - when a single cert asks to cover the `domain.tld` and all subdomains `*.domain.tld` ... which is probably fairly common since many sites want www.d.t and naked d.t to refer to the same web site.

**Comment by josemaia** [comment]
Hello guys, when can I expect there to be a new release of sewer containing the powerDNS provider? I implemented this provider manually in my project using sewer but I'd rather have a better-tested version!

**Comment by mmaney** [comment]
Have you been following the master branch here @josemaia?  I've been working on reducing my "future" branch to provide the changes that still need to go in for the next 0.8 release, and I hope to have a PR (perhaps lacking documentation updates that also need to go in) later today.  That will have a couple of the recent DNS providers as well as the protocol updates, and should be worthy of your attention.

And after that I expect 0.8 will get bugfixes, maybe some new provider modules, but not the less compatible changes I've been working on in my local tree.  That stuff will go into a 0.9 branch that will be more experimental for a while until it's ready for release.

Oops, almost lost my question for you: you're hoping for a better-tested what, exactly?  Since the protocol changes went in the ACME code in client has been able to be exercised against LE's staging and has been tested that way.  And getting that working again since they dropped support for the earlier draft protocol in staging, hence being ready for production to drop it later this year, has been one of my goals this year.  As for the powerdns driver, or most any of the providers, the only real-world testing they can get is from people like you, who have access to the service provider's API.  Rejoice, for powerdns has at least two such users now (that I know of)!  :-)

**Comment by josemaia** [comment]
By better-tested I simply meant using Kyle's code instead of my own homebrewed version, haha. Thank you for the update, looking forward to it.

**Pull request merged**